worker/rhpv3.go

```diff
@@ -837,6 +837,191 @@ func RPCReadRegistry(t *rhpv3.Transport, payment rhpv3.PaymentMethod, key rhpv3.
+
+	// Starting from here, we need to make sure to release the txn on error.
+	defer func() {
+		if err != nil {
+			ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+			defer cancel()
+			discardErr := w.bus.WalletDiscard(ctx, wprr.TransactionSet[len(wprr.TransactionSet)-1])
+			if discardErr != nil {
+				w.logger.Errorf("failed to discard txn after failed renewal: %v", discardErr)
+			}
+		}
+	}()
+
+	txnSet := wprr.TransactionSet
+	parents, txn := txnSet[:len(txnSet)-1], txnSet[len(txnSet)-1]
+
+	// Sign only the revision and contract. We can't sign everything because
+	// then the host can't add its own outputs.
+	h := types.NewHasher()
+	txn.FileContracts[0].EncodeTo(h.E)
+	txn.FileContractRevisions[0].EncodeTo(h.E)
+	finalRevisionSignature := renterKey.SignHash(h.Sum())
+
+	// Send the request.
+	req := rhpv3.RPCRenewContractRequest{
+		TransactionSet:         txnSet,
+		RenterKey:              rev.UnlockConditions.PublicKeys[0],
+		FinalRevisionSignature: finalRevisionSignature,
+	}
+	if err = s.WriteResponse(&req); err != nil {
+		return rhpv2.ContractRevision{}, nil, err
+	}
+
+	// Incorporate the host's additions.
+	var hostAdditions rhpv3.RPCRenewContractHostAdditions
+	if err = s.ReadResponse(&hostAdditions, 4096); err != nil {
+		return rhpv2.ContractRevision{}, nil, err
+	}
+	parents = append(parents, hostAdditions.Parents...)
+	txn.SiacoinInputs = append(txn.SiacoinInputs, hostAdditions.SiacoinInputs...)
+	txn.SiacoinOutputs = append(txn.SiacoinOutputs, hostAdditions.SiacoinOutputs...)
+	finalRevRenterSig := types.TransactionSignature{
+		ParentID:       types.Hash256(rev.ParentID),
+		PublicKeyIndex: 0, // renter key is first
+		CoveredFields: types.CoveredFields{
+			FileContracts:         []uint64{0},
+			FileContractRevisions: []uint64{0},
+		},
+		Signature: finalRevisionSignature[:],
+	}
+	finalRevHostSig := types.TransactionSignature{
+		ParentID:       types.Hash256(rev.ParentID),
+		PublicKeyIndex: 1,
+		CoveredFields: types.CoveredFields{
+			FileContracts:         []uint64{0},
+			FileContractRevisions: []uint64{0},
+		},
+		Signature: hostAdditions.FinalRevisionSignature[:],
+	}
+	txn.Signatures = []types.TransactionSignature{finalRevRenterSig, finalRevHostSig}
+
+	// Sign the inputs we funded the txn with and cover the whole txn including
+	// the existing signatures.
+	cf := types.CoveredFields{
+		WholeTransaction: true,
+		Signatures:       []uint64{0, 1},
+	}
+	if err := w.bus.WalletSign(ctx, &txn, wprr.ToSign, cf); err != nil {
+		return rhpv2.ContractRevision{}, nil, err
+	}
+
+	// Create a new no-op revision and sign it.
+	noOpRevision := initialRevision(txn, rev.UnlockConditions.PublicKeys[1], renterKey.PublicKey().UnlockKey())
+	h = types.NewHasher()
+	noOpRevision.EncodeTo(h.E)
+	renterNoOpSig := renterKey.SignHash(h.Sum())
+	renterNoOpRevisionSignature := types.TransactionSignature{
+		ParentID:       types.Hash256(noOpRevision.ParentID),
+		PublicKeyIndex: 0, // renter key is first
```

Comment:
might be silly but I think `RenterIndex` and `HostIndex` constants might make some of this transaction stuff easier to read from time to time. I know renter is first but I always have to check to make 100% sure.

Reply:
I agree. I'd put that into core first though since we might as well update all 3 packages in `renterd` to use that constant instead of only the worker.